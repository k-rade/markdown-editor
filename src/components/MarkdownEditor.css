.markdown-editor {
  height: 100vh;
  margin: auto;
  display: flex;
}

.header {
  box-shadow: -1px 0px 0 var(--color-black-two) inset;
  margin: 0 0 0;
  padding: 1.25rem;
  height: 100vh;
  z-index: 1;
  top: 0;
  left: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 1.25rem;
  background-color: var(--color-gray-900);
  border-right: 1px solid var(--color-gray-500);
}

.logo {
  margin: 0.5rem 0 0;
  font-weight: 600;
  font-size: 1.25rem;
}

.tabs {
  gap: 1rem;
  display: flex;
  flex-direction: column;
}

.tab {
  cursor: pointer;
  font-size: 1.25rem;
  font-weight: 600;
  margin: 0;
  position: relative;
  color: var(--color-gray-200);
}

.tab-active {
  color: var(--color-white);
}

.tab-active::before {
  position: absolute;
  content: "";
  width: 4px;
  height: 3rem;
  top: -0.275rem;
  left: -1.25rem;
  border-top-right-radius: 8px;
  border-bottom-right-radius: 8px;
  background-color: var(--color-primary-500);
}

.actions {
  gap: 0.5rem;
  display: flex;
  flex-direction: column;
}

.output {
  overflow-y: auto;
  width: 100%;
  padding: 0 4rem;
}

.code-mirror {
  margin-top: 4rem;
  position: relative;
}

.code-mirror>div {
  height: 100%;
  background-color: transparent !important;
  font-family: var(--font);
  color: var(--color-white-two) !important;
}

.CodeMirror-selected {
  background-color: var(--color-gray-400) !important;
}

.CodeMirror-cursor {
  border-left: 2px solid var(--color-primary-500) !important;
}

.cm-hr {
  color: var(--color-primary-500);
}

.cm-md-strong {
  color: var(--color-gray-200) !important;
}

.cm-url {
  color: var(--color-primary-500) !important;
}

.cm-header {
  color: var(--color-white) !important;
}

.cm-header-1,
h1 {
  font-size: 4rem;
}

.cm-header-2,
h2 {
  font-size: 3rem;
}

.cm-header-3,
h3 {
  font-size: 2rem;
}

.cm-header-4,
h4 {
  font-size: 1.75rem;
}

.cm-header-5,
h5 {
  font-size: 1.5rem;
}

.cm-heading-6,
h6 {
  font-size: 1.25rem;
}

.cm-strong,
.cm-em {
  color: var(--color-white);
}

.cm-md-heading-prefix {
  color: var(--color-gray-300) !important;
}


@media screen and (max-width:800px) {
  .header {
    margin: 0;
    border: 0;
    border-radius: 0;
    border-bottom: 1px solid var(--color-gray-500);
  }

  .code-mirror {
    padding: 0 2rem;
  }

}