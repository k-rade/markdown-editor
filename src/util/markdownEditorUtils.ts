import { MarkdownFile } from "../types/markdown";

export const downloadFile = (currentFile: MarkdownFile): void => {
  const element = document.createElement("a");
  const file = new Blob([currentFile.markdown], {
    type: "text/plain",
  });
  element.href = URL.createObjectURL(file);
  element.download = `${currentFile.name}.md`;
  document.body.appendChild(element); // Required for this to work in FireFox
  element.click();
};
