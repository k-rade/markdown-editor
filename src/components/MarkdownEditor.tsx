import { EffectCallback, useEffect, useState } from "react";
import MarkdownOutput from "./MarkdownOutput";
import Editor from "./Editor";
import Sidenav from "./Sidenav";
import { downloadFile } from "../util/markdownEditorUtils";
import { MarkdownFile } from "../types/markdown";
import Files from "./Files";
import FileSave from "./FileSave";

const MarkdownEditor = (): JSX.Element => {
  const [files, setFiles] = useState<MarkdownFile[]>([]);
  const [currentFile, setCurrentFile] = useState<MarkdownFile>({
    name: "Untitled",
    markdown: "",
  });
  const [activeTab, setActiveTab] = useState<"editor" | "preview">("editor");
  const [showFiles, setShowFiles] = useState(false);

  const handleShowFilesToggle = (): void => {
    setShowFiles((prevState) => !prevState);
  };

  useEffect((): ReturnType<EffectCallback> => {
    const savedFiles = handleReadSavedFilesFromLocalStorage();

    if (savedFiles.length) {
      setFiles(savedFiles);
      setCurrentFile({ ...savedFiles[0] });
    } else {
      handleReadIntroductionFile();
    }
  }, []);

  const handleReadSavedFilesFromLocalStorage = (): MarkdownFile[] => {
    return (
      Object.entries(localStorage)
        .filter((item) => item[0].includes("md:"))
        .map((storage) => {
          return {
            name: storage[0].substring(3, storage[0].length),
            markdown: storage[1],
          };
        }) || []
    );
  };

  const handleReadIntroductionFile = (): void => {
    //@ts-ignore
    import(`../text.md`).then((res) => {
      fetch(res.default)
        .then((res) => {
          return res.text();
        })
        .then((res) => {
          setCurrentFile({ ...currentFile, markdown: res });
        });
    });
  };

  const handleFileSelect = (name: string): void => {
    setCurrentFile({
      name,
      markdown: localStorage.getItem(`md:${name}`) || "",
    });
  };

  const handleFileDelete = (name: string): void => {
    const updatedFiles = [...files].filter((f) => f.name !== name);
    setFiles(updatedFiles);

    localStorage.removeItem(`md:${name}`);
  };

  const handleFileSave = (): void => {
    const fileAlreadyExists = files.some(
      (f: MarkdownFile) => f.name === currentFile.name
    );

    if (fileAlreadyExists) {
      const updatedFiles = files.map((f: MarkdownFile) =>
        f.name === currentFile.name
          ? { ...f, markdown: currentFile.markdown }
          : f
      );
      setFiles(updatedFiles);
    } else {
      setFiles([
        ...files,
        { name: currentFile.name, markdown: currentFile.markdown },
      ]);
      setCurrentFile({ ...currentFile, name: currentFile.name });
    }

    localStorage.setItem(`md:${currentFile.name}`, currentFile.markdown);
  };

  const handleFileNameChange = (name: string): void => {
    setCurrentFile({ ...currentFile, name });
  };

  const handleFileContentClear = (): void => {
    setCurrentFile({ ...currentFile, markdown: "" });
  };

  return (
    <div className="flex-col flex bg-zinc-50 dark:bg-zinc-900/50 dark:bg-opacity-40 sm:flex-row h-full">
      <Sidenav
        activeTab={activeTab}
        changeTab={(tab) => setActiveTab(tab)}
        clearCurrentFile={handleFileContentClear}
        downloadFile={() => downloadFile(currentFile)}
        toggleShowFiles={handleShowFilesToggle}
      />
      <Files
        files={files}
        selectFile={handleFileSelect}
        deleteFile={handleFileDelete}
        showFiles={showFiles}
        toggleFilesShow={handleShowFilesToggle}
        currentFile={currentFile}
      />

      <div className="w-full">
        <FileSave
          fileName={currentFile.name}
          saveFile={handleFileSave}
          changeFileName={handleFileNameChange}
        />

        {activeTab === "editor" ? (
          <Editor file={currentFile} updateCurrentFile={setCurrentFile} />
        ) : (
          <MarkdownOutput markdown={currentFile.markdown} />
        )}
      </div>
    </div>
  );
};

export default MarkdownEditor;
