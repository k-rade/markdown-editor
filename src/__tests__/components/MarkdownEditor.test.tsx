import { fireEvent, render } from "@testing-library/react";
import MarkdownEditor from "../../components/MarkdownEditor";
import { downloadFile } from "../../util/markdownEditorUtils";

jest.mock("react-markdown", () => (props: any): JSX.Element => {
  return <div>{props.children}</div>;
});
jest.mock("remark-gfm", () => () => {});
jest.mock("react-syntax-highlighter/dist/esm/styles/prism", () => () => {});
jest.mock("../../util/markdownEditorUtils", () => ({
  downloadFile: jest.fn(),
}));

describe("MarkdownEditor", (): void => {
  afterEach((): void => {
    jest.clearAllMocks();
  });

  it("should render", (): void => {
    const { baseElement } = render(<MarkdownEditor />);

    expect(baseElement).toMatchSnapshot();
  });

  it("should render with output", (): void => {
    const { baseElement, getByTestId } = render(<MarkdownEditor />);

    fireEvent.click(getByTestId("eye-icon"));

    expect(baseElement).toMatchSnapshot();
  });

  it("should save file", (): void => {
    Storage.prototype.setItem = jest.fn();
    const { getByTestId, getByDisplayValue } = render(<MarkdownEditor />);

    fireEvent.change(getByDisplayValue("Untitled"), {
      target: { value: "New file" },
    });
    fireEvent.click(getByTestId("save-icon"));

    expect(localStorage.setItem).toHaveBeenNthCalledWith(1, "md:New file", "");
  });

  it("should delete file", (): void => {
    Storage.prototype.setItem = jest.fn();
    Storage.prototype.removeItem = jest.fn();
    const { getByTestId, getByDisplayValue, getByText } = render(
      <MarkdownEditor />
    );

    fireEvent.change(getByDisplayValue("Untitled"), {
      target: { value: "New file" },
    });
    fireEvent.click(getByTestId("save-icon"));
    expect(localStorage.setItem).toHaveBeenNthCalledWith(1, "md:New file", "");

    fireEvent.click(getByTestId("trash-icon"));
    fireEvent.click(getByText("Delete"));
    expect(localStorage.removeItem).toHaveBeenNthCalledWith(1, "md:New file");
  });

  it("should toggle files menu", (): void => {
    const { getByTestId, queryByTestId } = render(<MarkdownEditor />);

    fireEvent.click(getByTestId("folder-icon"));
    expect(getByTestId("files-overlay")).toBeInTheDocument();

    fireEvent.click(getByTestId("x-icon"));
    expect(queryByTestId("files-overlay")).not.toBeInTheDocument();
  });

  it("should download file", (): void => {
    Object.defineProperty(window, "localStorage", {
      value: {
        key: "md:something",
        value: "markdown",
      },
    });
    (downloadFile as jest.Mock).mockReturnValue("file");
    const { getByTestId } = render(<MarkdownEditor />);

    fireEvent.click(getByTestId("download-icon"));

    expect(downloadFile).toHaveBeenNthCalledWith(1, {
      name: "Untitled",
      markdown: "",
    });
  });
});
