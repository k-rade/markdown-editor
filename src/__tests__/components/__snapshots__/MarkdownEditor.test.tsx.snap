// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownEditor should render 1`] = `
<body>
  <div>
    <div
      class="flex-col flex bg-zinc-50 dark:bg-zinc-800 dark:bg-opacity-40 sm:flex-row h-full"
    >
      <div
        class="overflow-y-auto w-full sm:max-w-[5rem] bg-zinc-800 dark:bg-zinc-900 flex flex-col items-center gap-6"
      >
        <div
          class="p-4 justify-between flex items-center sm:justify-center sm:px-0 bg-emerald-600 dark:bg-emerald-600 w-full z-10"
        >
          <div
            class="w-8 h-8 border-white border-8 rounded-full"
          />
          <svg
            class="stroke-white sm:hidden"
            data-testid="menu-icon"
            fill="none"
            height="1.25rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.25rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="3"
              x2="21"
              y1="12"
              y2="12"
            />
            <line
              x1="3"
              x2="21"
              y1="6"
              y2="6"
            />
            <line
              x1="3"
              x2="21"
              y1="18"
              y2="18"
            />
          </svg>
        </div>
        <div
          class="-top-full sm:h-full sm:bg-none z-10 transition-[top] duration-300 absolute sm:relative sm:mt-0 py-2 sm:py-0 sm:top-0 mt-16 bg-zinc-800 flex flex-col items-center gap-1 sm:gap-4 w-full px-2 sm:px-4"
          data-testid="sidenav-menu"
        >
          <div
            class="bg-zinc-600 hover:cursor-pointer hover:bg-zinc-600/50 bg-opacity-50 flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="edit-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Editor
            </span>
          </div>
          <div
            class=" hover:cursor-pointer hover:bg-zinc-600/50 bg-opacity-50 flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="eye-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
              />
              <circle
                cx="12"
                cy="12"
                r="3"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Preview
            </span>
          </div>
          <div
            class="hover:bg-opacity-50 hover:bg-zinc-600 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="download-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
              />
              <polyline
                points="7 10 12 15 17 10"
              />
              <line
                x1="12"
                x2="12"
                y1="15"
                y2="3"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Download
            </span>
          </div>
          <div
            class="hover:bg-opacity-50 hover:bg-zinc-600 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="x-circle-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="15"
                x2="9"
                y1="9"
                y2="15"
              />
              <line
                x1="9"
                x2="15"
                y1="9"
                y2="15"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Clear
            </span>
          </div>
          <div
            class="lg:hidden hover:bg-opacity-50 hover:bg-zinc-600 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="folder-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M22 19a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h5l2 3h9a2 2 0 0 1 2 2z"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Files
            </span>
          </div>
          <div
            class="hidden sm:block bg-zinc-500 h-px w-8"
          />
          <div
            class="hover:bg-opacity-50 hover:bg-zinc-600 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="fill-zinc-100"
              data-testid="theme-switch-icon"
              fill="currentColor"
              height="1.25rem"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 1a7 7 0 1 0 0 14A7 7 0 0 0 8 1zm0 13V2a6 6 0 1 1 0 12z"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Theme
            </span>
          </div>
        </div>
      </div>
      <div
        class="-left-full  overflow-y-auto z-10 h-full transition-[left] duration-300 absolute lg:relative lg:left-0 lg:ml-0 shadow-sm border-r bg-white dark:bg-zinc-700 dark:border-zinc-500 w-full max-w-sm"
      >
        <div
          class="border-b h-16 flex items-center justify-between px-4 text-lg font-semibold dark:text-zinc-200 dark:border-zinc-500"
        >
          <div>
            Notes
             
            <span
              class="text-sm text-zinc-400 font-normal ml-1"
            >
              0 Files
            </span>
          </div>
          <svg
            class="lg:hidden dark:stroke-zinc-400 stroke-zinc-500"
            data-testid="x-icon"
            fill="none"
            height="1.375rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.375rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </div>
        <div
          class="flex flex-col items-center justify-center gap-1 mt-8"
        >
          <svg
            class="stroke-1 stroke-zinc-300 mb-1 dark:stroke-zinc-400"
            fill="none"
            height="2.5rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="2.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z"
            />
            <polyline
              points="14 2 14 8 20 8"
            />
            <line
              x1="12"
              x2="12"
              y1="18"
              y2="12"
            />
            <line
              x1="9"
              x2="15"
              y1="15"
              y2="15"
            />
          </svg>
          <div
            class="text-zinc-700 font-semibold dark:text-zinc-100"
          >
            No notes saved
          </div>
          <div
            class="text-sm text-zinc-400"
          >
            Get started by saving your first note.
          </div>
        </div>
      </div>
      <div
        class="w-full"
      >
        <div
          class="bg-white w-full border-b h-16 flex items-center justify-center text-lg font-semibold dark:text-zinc-200 dark:border-zinc-500 dark:bg-zinc-700"
        >
          <form
            class="flex items-center justify-between gap-8 w-full max-w-5xl px-4 sm:px-8"
          >
            <input
              class="focus:outline-none bg-transparent w-full"
              placeholder="File name"
              required=""
              type="text"
              value="Introduction"
            />
            <button
              type="submit"
            >
              <svg
                class="dark:stroke-zinc-400 cursor-pointer hover:stroke-cyan-600 dark:hover:stroke-cyan-500"
                data-testid="save-icon"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z"
                />
                <polyline
                  points="17 21 17 13 7 13 7 21"
                />
                <polyline
                  points="7 3 7 8 15 8"
                />
              </svg>
            </button>
          </form>
        </div>
        <div
          class="w-full max-w-5xl m-auto sm:mt-8 sm:px-8 h-[calc(100vh-8rem)] sm:h-[calc(100vh-6rem)]"
        >
          <div
            class="react-codemirror2 code-mirror overflow-y-auto shadow-lg h-full rounded-md rounded-b-none"
          >
            <div
              class="CodeMirror cm-s-material CodeMirror-wrap CodeMirror-empty"
              translate="no"
            >
              <div
                style="overflow: hidden; position: relative; width: 3px; height: 0px;"
              >
                <textarea
                  autocapitalize="off"
                  autocorrect="off"
                  spellcheck="false"
                  style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; min-height: 1em; outline: none;"
                  tabindex="0"
                />
              </div>
              <div
                class="CodeMirror-vscrollbar"
                cm-not-content="true"
                tabindex="-1"
              >
                <div
                  style="min-width: 1px;"
                />
              </div>
              <div
                class="CodeMirror-hscrollbar"
                cm-not-content="true"
                tabindex="-1"
              >
                <div
                  style="height: 100%; min-height: 1px;"
                />
              </div>
              <div
                class="CodeMirror-scrollbar-filler"
                cm-not-content="true"
              />
              <div
                class="CodeMirror-gutter-filler"
                cm-not-content="true"
              />
              <div
                class="CodeMirror-scroll"
                tabindex="-1"
              >
                <div
                  class="CodeMirror-sizer"
                  style="margin-left: 0px;"
                >
                  <div
                    style="position: relative;"
                  >
                    <div
                      class="CodeMirror-lines"
                      role="presentation"
                    >
                      <div
                        role="presentation"
                        style="position: relative; outline: none;"
                      >
                        <pre
                          class="CodeMirror-placeholder CodeMirror-line-like"
                          style="height: 0px; overflow: visible; direction: ltr;"
                        >
                          Start typing
                        </pre>
                        <div
                          class="CodeMirror-measure"
                        >
                          <pre
                            class="CodeMirror-line-like"
                          >
                            <span>
                              xxxxxxxxxx
                            </span>
                          </pre>
                        </div>
                        <div
                          class="CodeMirror-measure"
                        />
                        <div
                          style="position: relative; z-index: 1;"
                        />
                        <div
                          class="CodeMirror-cursors"
                        />
                        <div
                          class="CodeMirror-code"
                          role="presentation"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  style="position: absolute; height: 50px; width: 1px;"
                />
                <div
                  class="CodeMirror-gutters"
                  style="display: none;"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`MarkdownEditor should render with output 1`] = `
<body>
  <div>
    <div
      class="flex-col flex bg-zinc-50 dark:bg-zinc-800 dark:bg-opacity-40 sm:flex-row h-full"
    >
      <div
        class="overflow-y-auto w-full sm:max-w-[5rem] bg-zinc-800 dark:bg-zinc-900 flex flex-col items-center gap-6"
      >
        <div
          class="p-4 justify-between flex items-center sm:justify-center sm:px-0 bg-emerald-600 dark:bg-emerald-600 w-full z-10"
        >
          <div
            class="w-8 h-8 border-white border-8 rounded-full"
          />
          <svg
            class="stroke-white sm:hidden"
            data-testid="menu-icon"
            fill="none"
            height="1.25rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.25rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="3"
              x2="21"
              y1="12"
              y2="12"
            />
            <line
              x1="3"
              x2="21"
              y1="6"
              y2="6"
            />
            <line
              x1="3"
              x2="21"
              y1="18"
              y2="18"
            />
          </svg>
        </div>
        <div
          class="-top-full sm:h-full sm:bg-none z-10 transition-[top] duration-300 absolute sm:relative sm:mt-0 py-2 sm:py-0 sm:top-0 mt-16 bg-zinc-800 flex flex-col items-center gap-1 sm:gap-4 w-full px-2 sm:px-4"
          data-testid="sidenav-menu"
        >
          <div
            class=" hover:cursor-pointer hover:bg-zinc-600/50 bg-opacity-50 flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="edit-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Editor
            </span>
          </div>
          <div
            class="bg-zinc-600 hover:cursor-pointer hover:bg-zinc-600/50 bg-opacity-50 flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="eye-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
              />
              <circle
                cx="12"
                cy="12"
                r="3"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Preview
            </span>
          </div>
          <div
            class="hover:bg-opacity-50 hover:bg-zinc-600 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="download-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
              />
              <polyline
                points="7 10 12 15 17 10"
              />
              <line
                x1="12"
                x2="12"
                y1="15"
                y2="3"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Download
            </span>
          </div>
          <div
            class="hover:bg-opacity-50 hover:bg-zinc-600 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="x-circle-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="15"
                x2="9"
                y1="9"
                y2="15"
              />
              <line
                x1="9"
                x2="15"
                y1="9"
                y2="15"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Clear
            </span>
          </div>
          <div
            class="lg:hidden hover:bg-opacity-50 hover:bg-zinc-600 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="stroke-zinc-100"
              data-testid="folder-icon"
              fill="none"
              height="1.25rem"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M22 19a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h5l2 3h9a2 2 0 0 1 2 2z"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Files
            </span>
          </div>
          <div
            class="hidden sm:block bg-zinc-500 h-px w-8"
          />
          <div
            class="hover:bg-opacity-50 hover:bg-zinc-600 hover:cursor-pointer flex gap-4 items-center justify-start sm:justify-center px-4 py-2 sm:px-0 sm:py-3 rounded-md w-full"
          >
            <svg
              class="fill-zinc-100"
              data-testid="theme-switch-icon"
              fill="currentColor"
              height="1.25rem"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1.25rem"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 1a7 7 0 1 0 0 14A7 7 0 0 0 8 1zm0 13V2a6 6 0 1 1 0 12z"
              />
            </svg>
            <span
              class="text-zinc-50 sm:hidden"
            >
              Theme
            </span>
          </div>
        </div>
      </div>
      <div
        class="-left-full  overflow-y-auto z-10 h-full transition-[left] duration-300 absolute lg:relative lg:left-0 lg:ml-0 shadow-sm border-r bg-white dark:bg-zinc-700 dark:border-zinc-500 w-full max-w-sm"
      >
        <div
          class="border-b h-16 flex items-center justify-between px-4 text-lg font-semibold dark:text-zinc-200 dark:border-zinc-500"
        >
          <div>
            Notes
             
            <span
              class="text-sm text-zinc-400 font-normal ml-1"
            >
              0 Files
            </span>
          </div>
          <svg
            class="lg:hidden dark:stroke-zinc-400 stroke-zinc-500"
            data-testid="x-icon"
            fill="none"
            height="1.375rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1.375rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </div>
        <div
          class="flex flex-col items-center justify-center gap-1 mt-8"
        >
          <svg
            class="stroke-1 stroke-zinc-300 mb-1 dark:stroke-zinc-400"
            fill="none"
            height="2.5rem"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="2.5rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z"
            />
            <polyline
              points="14 2 14 8 20 8"
            />
            <line
              x1="12"
              x2="12"
              y1="18"
              y2="12"
            />
            <line
              x1="9"
              x2="15"
              y1="15"
              y2="15"
            />
          </svg>
          <div
            class="text-zinc-700 font-semibold dark:text-zinc-100"
          >
            No notes saved
          </div>
          <div
            class="text-sm text-zinc-400"
          >
            Get started by saving your first note.
          </div>
        </div>
      </div>
      <div
        class="w-full"
      >
        <div
          class="bg-white w-full border-b h-16 flex items-center justify-center text-lg font-semibold dark:text-zinc-200 dark:border-zinc-500 dark:bg-zinc-700"
        >
          <form
            class="flex items-center justify-between gap-8 w-full max-w-5xl px-4 sm:px-8"
          >
            <input
              class="focus:outline-none bg-transparent w-full"
              placeholder="File name"
              required=""
              type="text"
              value="Introduction"
            />
            <button
              type="submit"
            >
              <svg
                class="dark:stroke-zinc-400 cursor-pointer hover:stroke-cyan-600 dark:hover:stroke-cyan-500"
                data-testid="save-icon"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h11l5 5v11a2 2 0 0 1-2 2z"
                />
                <polyline
                  points="17 21 17 13 7 13 7 21"
                />
                <polyline
                  points="7 3 7 8 15 8"
                />
              </svg>
            </button>
          </form>
        </div>
        <div
          class="w-full max-w-5xl m-auto sm:mt-8 sm:px-8 h-[calc(100vh-8rem)] sm:h-[calc(100vh-6rem)]"
        >
          <div
            class="markdown-output overflow-y-auto shadow-lg bg-white h-full p-4 sm:p-8 rounded-md dark:bg-zinc-700"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
